core: pathfinding: look for fastest path instead of shortest

eckter opened the pull request:
Fix https://github.com/OpenRailAssociation/osrd/issues/7058

Change summary:

1. Move code from `STDCMHeuristic` into a new class (anything related to estimating min MRSP time)
2. Add a pathfinding parameter to define a cost function
3. Set the cost function to use MRSP times
4. Remove heuristic implementation and tests (should be fixed before reintroduced)


Note: there's approximations made here. We don't actually run simulation and only use MRSPs. We also don't know the rolling stop speed or length as we'd need to add parameters to the request. Even if we add them, speed limits wouldn't carry over blocks as they should (to account for the RS length).

We're generally after approximate values there so I don't think it would cause any issue. 

I tested on a few trains on imported infras, path lengths go up and travel time go down.
Paris -> Miramas on one arbitrary rolling stock: 790km -> 808km, 12:52 -> 12:04

eckter committed:
core: pathfinding: look for fastest path instead of shortest

Erashin commented on core/src/test/java/fr/sncf/osrd/utils/graph/PathfindingTests.kt:
```diff
@@ -702,6 +707,48 @@ class PathfindingTests {
     }
 
+    /**
+     * Check that the cost function is used, shortest path is longest. This test use STDCM tooling
```
```suggestion
     * Check that the cost function is used, shortest path is longest. This test uses STDCM tooling
```

Erashin commented on core/src/main/kotlin/fr/sncf/osrd/utils/CachedBlockMRSPBuilder.kt:
```diff
@@ -0,0 +1,44 @@
+    private val mrspCache = mutableMapOf<BlockId, Envelope>()
+
+    // 320km/h as default value (global max speed in France)
+    private val rsMaxSpeed = rollingStock?.maxSpeed ?: (320.0 / 3.6)
```
Comment disappeared for some reason.
This could be a pb in the future for other countries. For example, germany's top train speed is 550hkm/h, which is a lot higher than France's. Not saying we could /should handle this right now, but we'll have to keep it in mind.

eckter replied:
I don't think German's train lines are equipped for trains to go faster than 300km/h.

Ideally we'd change the editoast -> core API to include some simulation data, but that seems a little overkill here. The whole process does a lot of approximations anyway.

Khoyo commented on core/src/main/kotlin/fr/sncf/osrd/api/pathfinding/RemainingDistanceEstimator.kt:
```diff
@@ -13,6 +13,13 @@ import fr.sncf.osrd.utils.units.Offset
 /**
  * This is a function object that estimates the remaining distance to the closest target point,
  * using geo data. It is used as heuristic for A*.
+ *
+ * **LEGACY**
+ *
+ * Not currently in used as the geodesic computations are too expensive to compute, resulting in a
+ * net performance loss.
+ *
+ * TODO: before reusing this class, it now needs to return a minimum *time* instead of a distance.
```
I don't really this the value in keeping this class anymore, shouldn't we just delete it?

eckter replied:
It's supposed to be fixed and used soon-ish. But that's the second comment on this, so I'll just remove it and link the removal in the relevant issue.

Khoyo reviewed (approved):
Otherwise LGTM

SharglutDev reviewed (approved):
Lgtm (not tested)

eckter committed:
core: pathfinding: remove heuristic

The implementation was too slow to be worth using,
and the last change breaks it anyway
